proposals/accepted/051-ghc-base-libraries.rst

```diff
@@ -0,0 +1,330 @@
+
+The two goals are in risk of conflict, and we have seen such conflict in the past.
+The purpose of this proposal is to reconcile these goals, in a way that satisfies all parties.
+
+Each of us places different emphasis on these goals, but respects them both.
+The CLC does not want to constrain GHC unnecessarily; and the GHC team does not want to undermine CLC's efforts, indeed quite the reverse!
+
+The proposal is based on `HF Proposal: standard library reform <https://github.com/haskellfoundation/tech-proposals/pull/47>`__, but is independent of it.
+
+Things we all agree about
+=========================
+
+Here are some points that everyone agrees about:
+
+1. ``base``, and other packages that come with GHC, should adhere rigorously to the PVP.
+
+2. Any complicated package, certainly including ``base``, has implementation "internals" that it may want to expose to hard-core users, but not to regular clients.
+   The right pattern for accommodating this is described in `Nikita's blog post <https://nikita-volkov.github.io/internal-convention-is-a-mistake/>`__: have two packages, an "internals" one exposing the internals, and a "stable" one that exposes the stable API. Both adhere to the PVP.
+
+3. We should use this pattern for ``base``.
+
+4. The Core Libraries Committee curates the API of ``base`` (here is `the charter <https://github.com/haskell/core-libraries-committee#base-package>`__), including:
+
+   - Types, specifically including what instances are exposed
+
+   - Semantics (including strictness)
+
+   - Performance
+
+   - Semantic changes to documentation (the charter says *"Documentation changes normally fall under GHC developers purview, except significant ones (e.g., adding or changing type class laws)."*)
+
+   Beyond that, it has no interest in the implementation details (e.g. alpha renaming, moving things between modules, comments).
+
+5. In curating the ``base`` API, it is immaterial where code lives.
+   For example, if a change to ``ghc-prim`` changes the ``base`` API (as defined above) the GHC developers must consult the CLC.
+   The fact that the change isn't physically part of the ``base`` package is immaterial.
+   (Incidentally, ``base`` and ``ghc-prim`` are both part of the same GitHub repository, which also includes GHC's source code.)
+
+Proposal
+========
+
+We propose to divide ``base`` into three packages:
+
+- ``ghc-internals``: exposes aspects of GHC's internals that may be of interest to "hard-core" developers interested in maximum performance (see `Nikita's blog post <https://nikita-volkov.github.io/internal-convention-is-a-mistake/>`__).
+  The API of ``ghc-internals`` is fully under the control of the GHC team, and of no direct interest to the CLC — only its effects on the API of base.
+
+- ``base``: as now, whose API is curated by CLC.
+  Depends on ``ghc-internals``, and hence on ``ghc-bignum`` and ``ghc-prim``.
+
+- ``ghc-experimental``, initially empty, depends on ``base`` and on ``ghc-internals``.
```

Comment:
OK, so the governance structure is:

- `base` governed by CLC and CLC proposals
- `ghc-internals` governed solely by GHC team, no proposal necessary
- `ghc-experimental` governed by GHC proposal process

The Win here is that GHC Team can unilaterally make changes to `ghc-internals` without asking anyone, which allows them to iterate quickly. GHC Proposals can go in `ghc-experimental` without involving CLC directly. And CLC has less extraneous work. Feels like a win-win-win to me.